### PR TITLE
internal: Benchmark CI

### PR DESCRIPTION
**Pull request by ntucker**
### Motivation

Track performance improvements and regressions

### Solution

https://github.com/rhysd/github-action-benchmark

### A note

As far as watching the benchmark results of examples in this repository, the amplitude of the benchmarks is about +- 10~20%. If your benchmarks use some resources such as networks or file I/O, the amplitude might be bigger.

If the amplitude is not acceptable, please prepare a stable environment to run benchmarks. GitHub action supports self-hosted runners.

### CODE REVIEW

File: .github/workflows/benchmark.yml [hunk 1]
```diff
@@ -0,0 +1,55 @@
+name: Benchmark
+
+on:
+  pull_request:
+  push:
+    branches:
+      - master
```
Comment (ljharb) [inline .github/workflows/benchmark.yml:7]: why not on every branch? there's no allotment on actions usage.

**Review by ljharb — APPROVED** [review]
i have never seen a benchmarking approach that provides a consistent baseline; if this does, i'd love to figure out how you did it and use it elsewhere :-)

File: .github/workflows/benchmark.yml (hunk 1 above)
Comment (ntucker) [reply .github/workflows/benchmark.yml:7]: only interesting in tracking things against master - as both the comparison point and points on a graph.

PRs will still measure but not log the data

**Comment by ntucker** [comment]
Not sure what you mean about consistent baseline. Synthetic benches are easier to control; the challenge here is simply hardware variance - hence the note about only really to +/- 10% accuracy.

**Pull request merged**